cereal/messaging/msgq_to_zmq.cc

```diff
@@ -0,0 +1,133 @@
+#include "cereal/messaging/msgq_to_zmq.h"
+
+#include <cassert>
+#include <future>
+
+#include "common/util.h"
+
+extern ExitHandler do_exit;
+
+static std::string recv_zmq_msg(void *sock) {
+  zmq_msg_t msg;
+  zmq_msg_init(&msg);
+  std::string ret;
+  if (zmq_msg_recv(&msg, sock, 0) > 0) {
+    ret.assign((char *)zmq_msg_data(&msg), zmq_msg_size(&msg));
+  }
+  zmq_msg_close(&msg);
+  return ret;
+}
+
+void MsgqToZmq::run(const std::vector<std::string> &endpoints, const std::string &ip) {
+  zmq_context = std::make_unique<ZMQContext>();
+  msgq_context = std::make_unique<MSGQContext>();
+
+  // Create ZMQPubSockets for each endpoint
+  for (auto endpoint : endpoints) {
+    auto &socket_pair = socket_pairs.emplace_back();
+    socket_pair.endpoint = endpoint;
+    socket_pair.pub_sock = std::make_unique<ZMQPubSocket>();
+    int ret = socket_pair.pub_sock->connect(zmq_context.get(), endpoint);
+    if (ret != 0) {
+      printf("Failed to create ZMQ publisher for [%s]: %s\n", endpoint.c_str(), zmq_strerror(zmq_errno()));
+      return;
+    }
+  }
+
+  // Start ZMQ monitoring thread to monitor socket events
+  auto monitor_future = std::async(std::launch::async, &MsgqToZmq::zmqMonitorThread, this);
+  int n = 0;
+  while (!do_exit) {
+    std::unique_lock lk(mutex);
+    cv.wait(lk, [this]() { return do_exit || !sub2pub.empty(); });
+    if (do_exit) break;
+
+    printf("begin push %d\n", ++n);
+    for (auto sub_sock : msgq_poller->poll(100)) {
+      std::unique_ptr<Message> msg(sub_sock->receive(true));
+      while (msg && sub2pub[sub_sock]->sendMessage(msg.get()) == -1) {
+        if (errno != EINTR) break;
+      }
+    }
+  }
+}
+
+void MsgqToZmq::zmqMonitorThread() {
+  std::vector<zmq_pollitem_t> pollitems;
+
+  // Set up ZMQ monitor for each pub socket
+  for (int i = 0; i < socket_pairs.size(); ++i) {
+    std::string addr = "inproc://op-bridge-monitor-" + std::to_string(i);
+    zmq_socket_monitor(socket_pairs[i].pub_sock->sock, addr.c_str(), ZMQ_EVENT_ACCEPTED | ZMQ_EVENT_DISCONNECTED);
+
+    void *monitor_socket = zmq_socket(zmq_context->getRawContext(), ZMQ_PAIR);
+    zmq_connect(monitor_socket, addr.c_str());
+    pollitems.emplace_back(zmq_pollitem_t{.socket = monitor_socket, .events = ZMQ_POLLIN});
+  }
+
+  int n = 0;
+  while (!do_exit) {
+    int ret = zmq_poll(pollitems.data(), pollitems.size(), 1000);
+    if (ret < 0) {
+      if (errno == EINTR) {
+        // Due to frequent EINTR signals from msgq, introduce a brief delay (200 ms)
+        // to reduce CPU usage during retry attempts.
+        util::sleep_for(200);
```

Comment:
We need to add a sleep here to reduce CPU usage while polling. 
the MSGQ uses SIGUSR1 for notification between publishers and subscribers, causing frequent EINTR signals. This interrupts the poll function frequently, leading to a busy retry loop. This can increase CPU usage from an expected 0% to around 5%. Other interrupt-driven modules like camerad may face similar issues.

https://github.com/commaai/msgq/pull/617 is a final solution. However, it still needs more testing to confirm its reliability.